spyder/plugins/editor/widgets/tests/test_bookmarks.py

```diff
@@ -0,0 +1,275 @@
+    cursor.movePosition(QTextCursor.Right, n=2)
+    editor.setTextCursor(cursor)
+    ab(2)
+    editor_assert_helper(editor, block, bm=[(1, 0), (2, 2)], emits=True)
+
+
+def test_get_bookmarks(code_editor_bot):
+    """Test CodeEditor.get_bookmarks. Returns data found in textblocks."""
+    editor, _ = code_editor_bot
+    gb = editor.get_bookmarks
+
+    assert(gb() == {})
+
+    # Add bookmarks.
+    bm = {1: ('filename', 1, 0), 2: ('filename', 3, 5), 3: ('filename', 4, 3)}
+    editor.set_bookmarks(bm)
+    assert(gb() == {1: [1, 0], 2: [3, 5], 3: [4, 3]})
+
+
+def test_clear_bookmarks(code_editor_bot):
+    """Test CodeEditor.clear_bookmarks. Removes bookmarks from all blocks."""
+    editor, _ = code_editor_bot
+
+    assert len(editor.blockuserdata_list) == 0
+
+    bm = {1: ('filename', 1, 0), 2: ('filename', 3, 5)}
+    editor.set_bookmarks(bm)
+    assert editor.get_bookmarks() == {1: [1, 0], 2: [3, 5]}
+    assert len(editor.blockuserdata_list) == 2
+
+    editor.clear_bookmarks()
+    assert editor.get_bookmarks() == {}
+    # Even though there is a 'del data' that would pop the item from the
+    # list, the __del__ funcion isn't called.
+    assert len(editor.blockuserdata_list) == 2
+    for data in editor.blockuserdata_list:
+        assert not data.bookmarks
+
+
+def test_update_bookmarks(code_editor_bot):
+    """Test CodeEditor.update_bookmarks. Check if signal is emitted."""
+    editor, _ = code_editor_bot
+    reset_emits(editor)
+    editor.bookmarks_changed.emit.assert_not_called()
+    # update_bookmarks is the slot for the blockCountChanged signal.
+    editor.textCursor().insertBlock()
+    editor.bookmarks_changed.emit.assert_called_with()
+
+
+def test_save_bookmark(setup_editor):
+    """Test Plugin.save_bookmark. 
+
+    Test saving of bookmarks by looking at data in blocks. Reassignment
+    should remove data from old block and put it in new.
+    """
+    editor = setup_editor
+
+    editor.setup_open_files()
+
+    # Get current editorstack, active editor and cursor
+    editorstack = editor.get_current_editorstack()
+    edtr = editorstack.get_current_editor()
+    cursor = edtr.textCursor()
+
+    # Basic functionality: save a bookmark and check if it's there
+    editor.save_bookmark(1)
+    bookmarks = edtr.document().findBlockByNumber(0).userData().bookmarks
+    assert bookmarks == [(1, 0)]
+
+    # Move the cursor and reset the same bookmark
+    cursor.movePosition(QTextCursor.Down, n=1)
+    cursor.movePosition(QTextCursor.Right, n=2)
+    edtr.setTextCursor(cursor)
+    editor.save_bookmark(1)
+
+    # Check if bookmark is there
+    bookmarks = edtr.document().findBlockByNumber(1).userData().bookmarks
+    assert bookmarks == [(1, 2)]
+
+    # Check if bookmark was removed from previous block
+    bookmarks = edtr.document().findBlockByNumber(0).userData().bookmarks
+    assert bookmarks == []
+
+
+def test_load_bookmark(setup_editor, test_file):
+    """Test that loading a bookmark works.
+
```

Comment:
Our agreed convention is to make the summary line on its own line for multi-line docstrings